pilot/pkg/proxy/envoy/v2/debounced_channel.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"sync"
+	"time"
+)
+
+// DebouncedEvent is used as entry type, which is enqueued in DebouncedChannel
+type DebouncedEvent = bool
+
+// DebouncedChannel is a channel which abstracts the debounce algorithm
+type DebouncedChannel struct {
+	// C channel, which can be used in selects. It delivers debounced events.
+	// The function emitted must be called. Otherwise processing is stopped.
+	C <-chan func(func(DebouncedEvent))
+	c chan func(func(DebouncedEvent))
+
+	pushCounter      uint32
+	debounceCounter  uint32
+	debounceAfterEnd time.Time
+	debounceMaxEnd   time.Time
+	lastEnqueueTime  time.Time
+	lastEmitTime     time.Time
+	debounceAfter    time.Duration
+	debounceMax      time.Duration
+	mutex            sync.Mutex
+
+	merge              func(accumulator DebouncedEvent, value DebouncedEvent) DebouncedEvent
+	accumulator        DebouncedEvent
+	initialAccumulator DebouncedEvent
+
+	running bool
+}
+
+// NewDebouncedChannel create a new instance of DebouncedChannel
+// debounceAfter: A new outgoing event is emitted if no additional event is enqueued during this period (resets on every enqueued event).
+// debounceMax: A new outgoing event is emitted after this period (even if additional events were enqueued in that period).
+// initialAccumulator: Initial value of the accumulator
+// merge: function, which is called to merge two events if they are debounced.
+func NewDebouncedChannel(debounceAfter time.Duration, debounceMax time.Duration,
+	initialAccumulator DebouncedEvent, merge func(accumulator DebouncedEvent, value DebouncedEvent) DebouncedEvent) *DebouncedChannel {
+
+	c := make(chan func(func(DebouncedEvent)), 1)
```

Comment:
Will this block the calling of `Enqueue` as previously we have `updateChannel` of 10 buffer?
If this can block, then for example when a large number of pods/endpoints update,  the `ConfigUpdate` will also block. If so, this will cause snowslide.

So basically i think this is of risk

Reply:
No, this can never block. The logic in `runIfNeeded` avoids that more than one entry is written to this channel.

Reply:
See also the test `TestDebouncedChannelMultiple`. If `Enqueue` would block, this test will never come to an end.